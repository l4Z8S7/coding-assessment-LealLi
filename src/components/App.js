import React, { Component } from 'react';
import './App.css';
import Welcome from './welcome/welcome';
import Card from './card/card';
import Review from './review/review';
import Result from './result/result';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      questions: [],
      answers: [],
      current_index: -1,
      submitted: false
    }
    this.selectAnswer = this.selectAnswer.bind(this);
  }

  componentDidMount() {
    fetch('questions.json')
    .then(res => res.json())
    .then(data => {
      this.setState({ questions: data, answers: new Array(data.length).fill("") });
    });
  }

  selectAnswer(answer) {
    const answers = [...this.state.answers];
    if (answers[this.state.current_index] !== answer) {
      answers[this.state.current_index] = answer;
      this.setState({ answers });
    }
  }

  _caculateScore() {
    return this.state.answers.map((answer, index) => {
      if (answer) {
        return this.state.questions[index].options[answer];
      } else {
        return 0;
      }
    }).reduce((a, b) => a + b, 0);
  }

  renderContent(index) {
    if (this.state.submitted) {
      return <Result score={this._caculateScore()}/>
    }
    const total_index = this.state.questions.length;
    if (index === -1) {
      return <Welcome startAs={() => this.setState({ current_index: 0 })}/>;
    } else if (index === total_index) {
      return <Review 
        answers={this.state.answers} 
        returnTo={index => this.setState({ current_index: index })}
        submitAs={() => this.setState({ submitted: true })}
      />;
    } else {
      return <Card
        question={this.state.questions[index]}
        answer={this.state.answers[index]}
        prevQs={() => this.setState({ current_index: this.state.current_index - 1 })}
        nextQs={() => this.setState({ current_index: this.state.current_index + 1 })}
        selectAs={this.selectAnswer}
        index={index}
      />;
    }
  }

  render() {
    return (
      <div className="App">
        { this.renderContent(this.state.current_index) }
      </div>
    );
  }
}

export default App;
