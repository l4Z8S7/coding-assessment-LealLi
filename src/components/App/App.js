import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import './App.css';
import Welcome from '../welcome/welcome';
import Card from '../card/card';
import Review from '../review/review';
import Result from '../result/result';
import * as actions from '../../actions';

export class App extends Component {
  componentDidMount() {
    this.props.fetchQuestions();
  }
  
  render() {
    return (
      <div className="App">
      {
        this.props.submitted ? <Result /> : 
        this.props.index === -1 ? <Welcome /> : 
        this.props.index === this.props.questions.length ? <Review /> : 
        <Card />
      }
      </div>
    );
  }
}

function mapStateToProps({ questions, index, submitted }) {
  return { questions, index, submitted };
}

App.propTypes = {
  questions: PropTypes.arrayOf(PropTypes.object).isRequired,
  index: PropTypes.number.isRequired,
  submitted: PropTypes.bool.isRequired,
  fetchQuestions: PropTypes.func.isRequired
}

export default connect(mapStateToProps, actions)(App);
