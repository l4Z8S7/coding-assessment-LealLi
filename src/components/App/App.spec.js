import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { App } from './App';
import Welcome from '../welcome/welcome';
import Card from '../card/card';
import Review from '../review/review';
import Result from '../result/result';

Enzyme.configure({ adapter: new Adapter });

const props = {
  questions: [{ title: 'Q1', options: { 'Opt1': 1, 'Opt2': 2} }],
  index: -1,
  submitted: false,
  fetchQuestions: () => {}
}

describe('App', () => {
  it('should render Welcome component when index is -1', () => {
    expect(toJson(shallow(<App {...props} />))).toMatchSnapshot();
    expect(shallow(<App {...props} />).contains(<Welcome />)).toBe(true);
  });

  it('should render Card component when index is the number between 0 and length of questions', () => {
    expect(toJson(shallow(<App {...props} index={0} />))).toMatchSnapshot();
    expect(shallow(<App {...props} index={0} />).contains(<Card />)).toBe(true);
  });

  it('should render Review component when navigate to the end', () => {
    expect(toJson(shallow(<App {...props} index={1} />))).toMatchSnapshot();
    expect(shallow(<App {...props} index={1} />).contains(<Review />)).toBe(true);
  });

  it('should render Result component when submiited is true', () => {
    expect(toJson(shallow(<App {...props} submitted={true} />))).toMatchSnapshot();
    expect(shallow(<App {...props} submitted={true} />).contains(<Result />)).toBe(true);
  });
})