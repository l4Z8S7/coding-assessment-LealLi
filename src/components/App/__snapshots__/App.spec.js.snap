// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render Card component when index is the number between 0 and length of questions 1`] = `
<div
  className="App"
>
  <Connect(Component) />
</div>
`;

exports[`App should render Result component when submiited is true 1`] = `
<div
  className="App"
>
  <Connect(Component) />
</div>
`;

exports[`App should render Review component when navigate to the end 1`] = `
<div
  className="App"
>
  <Connect(Component) />
</div>
`;

exports[`App should render Welcome component when index is -1 1`] = `
<div
  className="App"
>
  <Connect(Component) />
</div>
`;
